File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day06ProbablyAFireHazard.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

abstract class Day06Command(top: Int, left: Int, bottom: Int, right: Int)

case class Day06TurnOnCommand(top: Int, left: Int, bottom: Int, right: Int) extends Day06Command(top, left, bottom, right)

case class Day06ToggleCommand(top: Int, left: Int, bottom: Int, right: Int) extends Day06Command(top, left, bottom, right)

case class Day06TurnOffCommand(top: Int, left: Int, bottom: Int, right: Int) extends Day06Command(top, left, bottom, right)


class Day06ProbablyAFireHazard(filename: String) extends DailyProblem[Int, Int] {
  private val turnOnPattern = "turn on ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)".r
  private val togglePattern = "toggle ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)".r
  private val turnOffPattern = "turn off ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)".r

  private def parser(line: String): Day06Command =
    line match {
      case turnOnPattern(t, l, b, r) => Day06TurnOnCommand(t.toInt, l.toInt, b.toInt, r.toInt)
      case togglePattern(t, l, b, r) => Day06ToggleCommand(t.toInt, l.toInt, b.toInt, r.toInt)
      case turnOffPattern(t, l, b, r) => Day06TurnOffCommand(t.toInt, l.toInt, b.toInt, r.toInt)
      case _ => throw new IllegalStateException(s"Cannot match '$line'")
    }

  private def executeCommands(commands: List[Day06Command], lights: Array[Array[Int]], on: Int => Int, toggle: Int => Int, off: Int => Int): Int = {
    def process(top: Int, left: Int, bottom: Int, right: Int, operation: Int => Int) = {
      for (col <- left to right)
        for (row <- top to bottom)
          lights(row)(col) = operation(lights(row)(col))
    }

    commands.foreach {
      case Day06TurnOnCommand(t, l, b, r) => process(t, l, b, r, on)
      case Day06ToggleCommand(t, l, b, r) => process(t, l, b, r, toggle)
      case Day06TurnOffCommand(t, l, b, r) => process(t, l, b, r, off)
    }

    lights.foldLeft(0)((acc, row) => acc + row.sum)
  }

  private val commands = FileUtilities.readFile(filename, parser)

  override val part1Answer = executeCommands(commands, Array.ofDim[Int](1000, 1000), _ => 1, a => if (a == 0) 1 else 0, _ => 0)
  override val part2Answer = executeCommands(commands, Array.ofDim[Int](1000, 1000), a => a + 1, a => a + 2, a => Math.max(0, a - 1))
}

File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day06ProbablyAFireHazard.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.twenty_fifteen.support.Light;
import advent.twenty_fifteen.support.LightCommand;
import advent.utilities.FileUtilities;

import java.util.List;

public class Day06ProbablyAFireHazard implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public Day06ProbablyAFireHazard(String filename) {
        List<LightCommand> commands = FileUtilities.readLines(filename, LightCommand::PARSE);

        Light[][] lights = new Light[1000][];
        for (int row = 0; row != 1000; row++) {
            lights[row] = new Light[1000];
            for (int col = 0; col != 1000; col++) {
                lights[row][col] = new Light();
            }
        }

        commands.forEach(command -> {
            for (int r = command.getStartRow(); r != command.getEndRow() + 1; r++) {
                for (int c = command.getStartCol(); c != command.getEndCol() + 1; c++) {
                    command.modify(lights[r][c]);
                }
            }
        });

        int nicePart1 = 0;
        int nicePart2 = 0;

        for (int row = 0; row != 1000; row++) {
            for (int col = 0; col != 1000; col++) {
                if (lights[row][col].isOn()) {
                    nicePart1++;
                }
                nicePart2 += lights[row][col].getBrightness();
            }
        }

        this.part1Answer = nicePart1;
        this.part2Answer = nicePart2;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}
